File: 백준/Gold/7579. 앱/앱.py
import sys
sys.setrecursionlimit(111111)
input = sys.stdin.readline

# 0-1 냅색 문제 인데 무게가 0인 물건 존재
# dp[i][j] = max(dp[i-1][j],dp[i-1][j-c[i]]+A[i])

n, m = map(int, input().split())
A = list(map(int, input().split()))
c = list(map(int, input().split()))




dp = [[0]*(sum(c)+1) for _ in range(n+1)]
result = sum(c)
for i in range(1,n+1):
        byte = A[i-1]
        cost = c[i-1]
        for j in range(sum(c)+1):
            if j<cost:
                  dp[i][j] = dp[i-1][j]
            else:
                dp[i][j] = max(dp[i-1][j],dp[i-1][j-c[i-1]]+A[i-1])
            if dp[i][j]>=m:
                  result=min(result,j)

if m !=0:
      print(result)
else:
      print(0)

File: 백준/Gold/7579. 앱/앱.java
import java.io.*;
import java.util.*;

public class Main {
	static final int MAX = (int) 1e9;

	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(in.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());

		int[] memory = new int[n];
		int[] cost = new int[n];
		int totalMem = 0;
		int totalCost = 0;

		st = new StringTokenizer(in.readLine());
		for (int i = 0; i < n; i++) {
			memory[i] = Integer.parseInt(st.nextToken());
			totalMem += memory[i];
		}

		st = new StringTokenizer(in.readLine());
		for (int i = 0; i < n; i++) {
			cost[i] = Integer.parseInt(st.nextToken());
			totalCost += cost[i];
		}

		int[] dp = new int[totalMem + 1];
		Arrays.fill(dp, MAX);
		dp[0] = 0;

		for (int i = 0; i < n; i++) {
			for (int j = totalMem; j >= memory[i]; j--) {
				if (dp[j - memory[i]] + cost[i] < dp[j]) {
					dp[j] = dp[j - memory[i]] + cost[i];
				}
			}
		}

		int result = MAX;
		for (int i = m; i <= totalMem; i++) {
			result = Math.min(result, dp[i]);
		}

		System.out.println(result);
	}
}
